rts/System/Sound/OpenAL/MusicStream.cpp

```diff
@@ -0,0 +1,316 @@
+/* This file is part of the Spring engine (GPL v2 or later), see LICENSE.html */
+#include "MusicStream.h"
+
+#include "System/FileSystem/FileHandler.h"
+#include "System/Sound/SoundLog.h"
+#include "System/SafeUtil.h"
+#include "ALShared.h"
+#include "VorbisShared.h"
+#include "System/Sound/OpenAL/OggDecoder.h"
+#include "System/Sound/OpenAL/Mp3Decoder.h"
+
+#include <cstring> //memset
+#include <filesystem>
+#include <string_view>
+#include <variant>
+
+// NOTE:
+//   this buffer gets recycled by each new stream, across *all* audio-channels
+//   as a result streams are limited to only ever being played within a single
+//   channel (currently BGMusic), but cause far less memory fragmentation
+// TODO:
+//   can easily be fixed if necessary by giving each channel its own index and
+//   passing that along to the callbacks via COggStream{::Play}
+// CFileHandler fileBuffers[NUM_AUDIO_CHANNELS];
+static CFileHandler fileBuffer("", "");
+
+// FIXME these parts of MusicStream are outside the class because every instance
+// of CSoundSource has a copy of MusicStream class but only one is ever active
+static constexpr unsigned int BUFFER_SIZE = 512 * 1024; // 512KB
+static std::variant<OggDecoder, Mp3Decoder> decoder;
+
+
+MusicStream::MusicStream(ALuint _source)
+	: pcmDecodeBuffer(nullptr)
+	, source(_source)
+	, format(AL_FORMAT_MONO16)
+	, stopped(true)
+	, paused(false)
+{
+	std::fill(buffers.begin(), buffers.end(), 0);
+}
+
+MusicStream::~MusicStream()
+{
+	Stop();
+	spring::SafeDeleteArray(pcmDecodeBuffer);
+}
+
+
+MusicStream& MusicStream::operator=(MusicStream&& rhs) noexcept
+{
+	if (this != &rhs) {
+		spring::SafeDeleteArray(pcmDecodeBuffer);
+		pcmDecodeBuffer = rhs.pcmDecodeBuffer;
+		rhs.pcmDecodeBuffer = nullptr;
+
+		for (auto i = 0; i < buffers.size(); ++i) {
+			std::swap(buffers[i], rhs.buffers[i]);
+		}
+
+		std::swap(source, rhs.source);
+		std::swap(format, rhs.format);
+
+		std::swap(stopped, rhs.stopped);
+		std::swap(paused, rhs.paused);
+
+		std::swap(msecsPlayed, rhs.msecsPlayed);
+		std::swap(lastTick, rhs.lastTick);
+	}
+
+	return *this;
+}
+
+// open an Ogg stream from a given file and start playing it
+void MusicStream::Play(const std::string& path, float volume)
+{
+	// we're already playing another stream
+	if (!stopped)
+		return;
+
+	fileBuffer.Close();
+	fileBuffer.Open(path);
+
+	if (!fileBuffer.FileExists()) {
+		LOG_L(L_ERROR, "[MusicStream::Play] File doesn't exist: %s", path.c_str());
+		return;
+	}
+
+	if (fileBuffer.GetFileExt() == std::string_view{"mp3"}) {
```

Comment:
does the `string_view` bring any benefit here over just `fileBuffer.GetFileExt() == "mp3"`?

Reply:
Defensive programming. I wanted to avoid a case where we compare raw const char* without strcmp. In this case GetFileExt is already a std::string but in case the return type changes the code without explicit string_view conversion would start failing.